hail/python/hail/expr/functions.py

```diff
@@ -6193,6 +6193,28 @@ def bit_not(x):
     return construct_expr(ir.ApplyUnaryPrimOp('~', x._ir), x.dtype, x._indices, x._aggregations)
 
 
+@typecheck(x=expr_oneof(expr_int32, expr_int64))
+def bit_count(x):
+    """Count the number of 1s in the binary representation of `x`.
+
+    Examples
+    --------
+    The binary representation of `7` is `111`, so:
+
+    >>> hl.eval(hl.bit_count(7))
+    3
+
+    Parameters
+    ----------
+    x : :class:`.Int32Expression` or :class:`.Int64Expression`
+
+    Returns
+    -------
+    :class:`.Int32Expression`
+    """
+    return construct_expr(ir.ApplyUnaryPrimOp('BitCount', x._ir), x.dtype, x._indices, x._aggregations)
```

Comment:
This should always return an int32 expression, not `x.dtype`.

It looks like you'll need to fix `ApplyUnaryPrimOp._compute_type` (in python) as well. Currently it assumes all unary ops preserve the type of the argument. Now you'll need to handle `BitCount` as a special case, and always return `tint32`.

Reply:
Updated!